core/serialization/blocks.js

```diff
@@ -263,8 +263,162 @@ const saveConnection = function(connection) {
+};
+
+/**
+ * Loads the block represented by the given state into the given workspace.
+ * This is defined internally so that the extra optional parameter doesn't
+ * clutter our external API.
+ * @param {!State} state The state of a block to deserialize into the workspace.
+ * @param {!Workspace} workspace The workspace to add the block to.
+ * @param {!Connection} parentConnection The optional parent connection to
+ *     attach the block to.
+ */
+const loadInternal = function(state, workspace, parentConnection = undefined) {
+  const block = workspace.newBlock(state['type'], state['id']);
+  loadCoords(block, state);
+  loadAttributes(block, state);
+  loadExtraState(block, state);
+  if (parentConnection &&
+      (block.outputConnection || block.previousConnection)) {
+    parentConnection.connect(
+        /** @type {!Connection} */
+        (block.outputConnection || block.previousConnection));
+  }
+  // loadIcons(block, state);
+  loadFields(block, state);
+  loadInputBlocks(block, state);
+  loadNextBlocks(block, state);
+};
+
+/**
+ * Applies any coordinate information available on the state object to the
+ * block.
+ * @param {!Block} block The block to set the position of.
+ * @param {!State} state The state object to reference.
+ */
+const loadCoords = function(block, state) {
+  const x = state['x'] === undefined ? 10 : state['x'];
+  const y = state['y'] === undefined ? 10 : state['y'];
+  block.moveBy(x, y);
+};
+
+/**
+ * Applies any attribute information available on the state object to the block.
+ * @param {!Block} block The block to set the attributes of.
+ * @param {!State} state The state object to reference.
+ */
+const loadAttributes = function(block, state) {
+  if (state['collapsed']) {
+    block.setCollapsed(true);
+  }
+  if (state['enabled'] === false) {
+    block.setEnabled(false);
+  }
+  if (state['editable'] === false) {
+    block.setEditable(false);
+  }
+  if (state['deletable'] === false) {
+    block.setDeletable(false);
+  }
+  if (state['movable'] === false) {
+    block.setMovable(false);
+  }
+  if (state['inline'] !== undefined) {
+    block.setInputsInline(state['inline']);
+  }
+  if (state['data'] !== undefined) {
+    block.data = state['data'];
+  }
+};
+
+/**
+ * Applies any extra state information available on the state object to the
+ * block.
+ * @param {!Block} block The block to set the extra state of.
+ * @param {!State} state The state object to reference.
+ */
+const loadExtraState = function(block, state) {
+  if (!state['extraState']) {
+    return;
+  }
+  block.loadExtraState(state['extraState']);
+};
+
+/**
+ * Applies any extra state information available on the state object to the
+ * block.
+ * @param {!Block} block The block to set the extra state of.
+ * @param {!State} state The state object to reference.
+ */
+const loadFields = function(block, state) {
+  if (!state['fields']) {
+    return;
+  }
+  for (const fieldName of Object.keys(state['fields'])) {
```

Comment:
For-of loops are a performance disaster in every browser.  Use a regular loop.  Here and once below.